presto-spi/src/main/java/com/facebook/presto/spi/PageBuilder.java

```diff
@@ -101,16 +108,33 @@ public void reset()
         }
     }
 
-    public Type getType(int channel)
+    public static PageBuilder resetNew(PageBuilder other)
     {
-        return types.get(channel);
+        requireNonNull(other, "other is null");
+        if (other.isEmpty()) {
+            return new PageBuilder(other.initialExpectedEntries,
+                    other.pageBuilderStatus.getMaxPageSizeInBytes(),
+                    other.types,
+                    Optional.of(other.blockBuilders));
+        }
+
+        BlockBuilder[] builders = new BlockBuilder[other.types.size()];
+        for (int i = 0; i < builders.length; i++) {
+            builders[i] = other.blockBuilders[i].resetNew(other.pageBuilderStatus.createBlockBuilderStatus());
```

Comment:
I'm pretty sure the `BlockBuilderStatus` must come from the `PageBuilderStatus` in the new block.